### ISSUE TITLE
Learning rate used for TransformerEmbeddings and SequenceTagger Architecture

### ISSUE THREAD

**Issue by igormis** [flairNLP/flair#1968]
HI, am using the SequenceTagger with the following params:
```
tagger: SequenceTagger = SequenceTagger(hidden_size=256,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type,
                                        use_crf=True)
```
Where for embeddings I am using stacked embeddings, composed of `FlairEmbeddings('multi-forward')`,
    `FlairEmbeddings('multi-backward')` and most importantly `TransformerWordEmbeddings('bert-base-multilingual-cased', pooling_operation='first_last',fine_tune=True, batch_size=8, layers='-1', use_scalar_mix=True, allow_long_sentences=True)
`
and then I train the model using:
```
trainer.train('resources/taggers/example-ner',
              train_with_dev=True,
              patience=2,
              max_epochs=100)
```
For this setup I have these questions (mainly connected with the **TransformerWordEmbeddings**)

1. I using the default learning rate for the FLAIR architecture (lr=0.1), whereas I also want to fine_tune the TransformerWordEmbeddings during training. Knowing that the BERT embeddings needs lower learning rate (2*10^-5) for fine tunning, how the LR is set for finetunning BERT embeddings, is some default LR used or the FLAIR learning rate of 0.1?
2. THe speed of training on HIGH-RAM GPU when executing the training on 600.000 rows in CONLL format is quite slow when fine_tune=True in the TransformerWordEmbeddings. Any idea how to increase the speed of training, achieving more or less the same accuracy (the RAM is not a problem, around 32 GB of RAM available on the GPU machine)?

Any help on this will be more than welcome.

**Comment by stefan-it** [comment]
Hi @igormis , when you want to use the fine-tuning feature of `TransformerWordEmbeddings` this should help:

```python
use_crf = False # set to True if you want CRF on-top
tagger: SequenceTagger = SequenceTagger(
                    hidden_size=256,
                    embeddings=embeddings,
                    tag_dictionary=tag_dictionary,
                    tag_type='ner',
                    use_crf=use_crf,
                    use_rnn=False,
                    reproject_embeddings=False,
                )
```

Initialize the trainer with `AdamW` optimizer:

```python
trainer = ModelTrainer(tagger, corpus, optimizer=torch.optim.AdamW)
```

Start with your initial learning and linearly decrease it to 0 (end of training) and a weight_decay of `0`. This is pretty much the setup as it is used for token classification example in Transformers library (see [here](https://github.com/huggingface/transformers/tree/master/examples/token-classification)):

```python
from torch.optim.lr_scheduler import OneCycleLR

trainer.train(f"resources/ner",
                    learning_rate=5.0e-5,
                    mini_batch_size=16,
                    mini_batch_chunk_size=1,
                    max_epochs=5,
                    scheduler=OneCycleLR,
                    weight_decay=0.,
                    train_with_dev=False,
                )
```

Hope this helps :hugs:

**Comment by igormis** [comment]
Tnx @stefan-it this should solve the question in 1). I will give it a try. Any idea how ti increase the speed-up in 2)

**Comment by stefan-it** [comment]
In my experimente with sequence labeling models and Flair, the bootleneck is always CPU. So e.g. if you use a server system (many cores, but less clock speed), training would be much slower than if you use a modern desktop cpu with clock speed >3ghz.

**Comment by igormis** [comment]
@stefan-it I am training the model using your example. I have one question about: _Start with your initial learning and linearly decrease it to 0 (end of training) and a weight_decay of 0_.
IN your example you are starting with` learning_rate=5.0e-5`. Can u please clarify how to start with initial learning rate of 0.1?

**Comment by igormis** [comment]
should I train the model using my setup and then again with your code?

**Comment by stefan-it** [comment]
Hi @igormis , then you can use `learniong_rate=0.1`. In my experiments, that include concatenation of Flair Embeddings with Transformer-based embeddings, I could always see a slightly worse performance - compared to using only Transformer-based embeddings. E.g. we showed that for historical texts [here](http://ceur-ws.org/Vol-2696/paper_173.pdf) (see table 6).

For Fine-Tuning Transformer-based models in combination with Adam, I recommend using a smaller learning rate (smaller than 0.1) :)

**Comment by igormis** [comment]
OK, just to be clear @stefan-it I am using following embeddings:
```
embedding_types: List[TokenEmbeddings] = [
    FlairEmbeddings('news-forward'),
    FlairEmbeddings('news-backward'),
    TransformerWordEmbeddings('roberta-large', pooling_operation='first_last',fine_tune=True, batch_size=8, layers='-1', use_scalar_mix=True, allow_long_sentences=True) 
]
```
For the training I used this snippet:
```
embeddings: StackedEmbeddings = StackedEmbeddings(embeddings=embedding_types)

# 5. initialize sequence tagger
from flair.models import SequenceTagger

tagger: SequenceTagger = SequenceTagger(hidden_size=256,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type,
                                        use_crf=True)

# 6. initialize trainer
from flair.trainers import ModelTrainer

trainer: ModelTrainer = ModelTrainer(tagger, corpus)

# 7. start training

trainer.train('resources/taggers/example-ner',
              train_with_dev=True,
              patience=1,
              max_epochs=150)
```
whereas you are suggesting to use:
```
embeddings: StackedEmbeddings = StackedEmbeddings(embeddings=embedding_types)

# 5. initialize sequence tagger
from flair.models import SequenceTagger

tagger: SequenceTagger = SequenceTagger(hidden_size=256,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type,
                                        use_crf=True)

# 6. initialize trainer
from flair.trainers import ModelTrainer
import torch
trainer: ModelTrainer = ModelTrainer(tagger, corpus, optimizer=torch.optim.AdamW)
from torch.optim.lr_scheduler import OneCycleLR
# 7. start training
trainer.train(f"resources/ner",
                    learning_rate=5.0e-5,

                    mini_batch_size=16,

                    mini_batch_chunk_size=1,

                    max_epochs=5,

                    scheduler=OneCycleLR,

                    weight_decay=0.,

                    train_with_dev=False,

                )
```
am I correct?

**Comment by stefan-it** [comment]
This looks good 👍 you should also try to enable/disable the CRF. In my experiments there's no need to use a CRF on-top when using the Fine-Tuning approach (in most experiments)!
